### ISSUE TITLE
Twisted should support "PEP 0492 -- Coroutines with async and await syntax"

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8088]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#8088|
|Type|enhancement|
|Created|2015-11-05 14:13:22Z|
|Branch|https://github.com/twisted/twisted/tree/await-8088|

https://www.python.org/dev/peps/pep-0492/

hynek mentions it's nice, and I'm skeptical -- but we should try and see if it works.

<details><summary>Searchable metadata</summary>

```
trac-id__8088 8088
type__enhancement enhancement
reporter__hawkowl hawkowl
priority__normal normal
milestone__Python_3_x Python-3.x
branch__await_8088 await-8088
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1446732802471575 1446732802471575
changetime__1470257415487791 1470257415487791
version__None None
owner__GitHub__noreply_____ GitHub <noreply@...>
cc__lukasaoz@... cc__slav0nic0@...
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_1"></a>|@Lukasa commented|
|-|-|

An example of this, using Dave Beazley's curio framework, can be found here: https://github.com/python-hyper/hyper-h2/blob/master/examples/curio/curio-server.py. That code is roughly analogous to the Twisted version you can find in the same repository.

What's a bit less clear to me is how this is going to work. I'm interested to see though!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl commented|
|-|-|

Okay, the tests pass, and it's only using minorly ugly hacks.

Please review, I think it's in a good state for some feedback.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban commented|
|-|-|

Harsh review ahead ... from a Python 2.7 only person :)

I don't understand what are the benefits of having this as part of Twisted core?

I think that we still support Python 2.7 , 3.3 and 3.4, so unless we get rid of supporting those versions we will never use this code in Twisted itself.

Besise using the 'async def` instead of `@decorator \n def` and `return` instead of `returnValue` are there any other benefits ? Can the code defined with `async def` and containing deferred be mixed with async io code? 

I think that it is better to have this as a side-project and not part of the Twisted core.

It feel strange to me that we are working to create code which works on both Python 2 and Python 3 and now we are introducing code which only works on Python3

Maybe we can have an external twisted3 project with functionalities available only in Python 3 ... but then some of them will only be supported on 3.5 and above :)

I don't have anything against having this code merged in Twisted core, but I am just wandering if having it as an external project would not be a better choice.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_7"></a>|@Lukasa commented|
|-|-|

Form my perspective, I want this in Twisted core because I'd like, as much as possible, for Twisted to integrate cleanly with asyncio. Twisted should not be making a case for itself as "asyncio you can use on Python 2", it should be making a case for itself as "a whole lot of powerful batteries that you can use on all versions of Python, including your asyncio application". That means that if it can feel just a little bit more native on asyncio, then it should do that.

Just because it isn't useful for developing Twisted itself doesn't mean it's not worth doing. =)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

I completely agree with lukasa and I think that this should definitely be in Twisted itself.  Or, at least, in `deferred` :).

For what it's worth, I feel like we _need_ this kind of functionality less to implement things within Twisted because it's mainly useful for orchestration of large, complex chunks, and we are often doing fine-grained construction of primitive building blocks, which don't benefit much from such high-level wrappers.  There are definitely places where it would be nice to have `await` as an independent keyword for clarity, but we aren't going to suffer much for it.  Whereas users' applications will often need to do this, and in fact may be doing it already with `asyncio` building blocks; they should be able to grab Twisted and slot it in off the shelf.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl commented|
|-|-|

PR: https://github.com/twisted/twisted/pull/453

**Comment by twisted-trac** [comment]
|[<img alt="1st1's avatar" src="https://avatars.githubusercontent.com/u/239003?s=50" width="50" height="50">](https://github.com/1st1)<a name="note_10"></a>|@1st1 commented|
|-|-|

I've looked at the PR -- the awaitable protocol implementation looks good.  I also like the `ensureDeferred` approach. Looks like this can be merged.

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|GitHub set owner to GitHub<br>GitHub set status to `closed`|
|-|-|

In changeset 57a23118ecd407c954a27e89d16a052064f27f78
```
#!CommitTicketReference repository="" revision="57a23118ecd407c954a27e89d16a052064f27f78"
Merge await-8088: Add PEP492 async/await support to Deferreds

Author: runciter, hawkowl
Reviewer: 1st1, glyph
Fixes: #8088
```
